File: 2532/solution.py
import heapq
import sys

from typing import List


class Solution:
    def findCrossingTime(self, n: int, k: int, time: List[List[int]]) -> int:
        wait_left = []
        wait_right = []
        work_left = []
        work_right = []

        for i, item in enumerate(time):
            # time[i] = [leftToRighti, pickOldi, rightToLefti, putNewi]
            heapq.heappush(wait_left, (-(item[0] + item[2]), -i))
        
        remain = n
        cur_time = 0
        
        while remain > 0 or len(work_right) or len(wait_right):
            while len(work_left) and work_left[0][0] <= cur_time:
                item = heapq.heappop(work_left)
                idx = item[1]
                heapq.heappush(wait_left, (-(time[idx][0] + time[idx][2]), -idx))
            while len(work_right) and work_right[0][0] <= cur_time:
                item = heapq.heappop(work_right)
                idx = item[1]
                heapq.heappush(wait_right, (-(time[idx][0] + time[idx][2]), -idx))
            if len(wait_right):
                item = heapq.heappop(wait_right)
                idx = -item[1]
                cur_time += time[idx][2]
                heapq.heappush(work_left, (cur_time + time[idx][3], idx))
            elif remain > 0 and len(wait_left):
                item = heapq.heappop(wait_left)
                idx = -item[1]
                cur_time += time[idx][0]
                heapq.heappush(work_right, (cur_time + time[idx][1], idx))
                remain -= 1
            else:
                tmp = sys.maxsize
                if len(work_left):
                    tmp = min(tmp, work_left[0][0])
                if len(work_right):
                    tmp = min(tmp, work_right[0][0])
                if tmp != sys.maxsize:
                    cur_time = max(cur_time, tmp)
        return cur_time

File: 2532/solution.go
package _2532

import "container/heap"

type waitItem struct{ totalTime, idx int }
type waitQueue []waitItem

func (w waitQueue) Len() int { return len(w) }
func (w waitQueue) Less(i, j int) bool {
	if w[i].totalTime == w[j].totalTime {
		return w[i].idx > w[j].idx
	} else {
		return w[i].totalTime > w[j].totalTime
	}
}
func (w waitQueue) Swap(i, j int)       { w[i], w[j] = w[j], w[i] }
func (w *waitQueue) Push(v interface{}) { *w = append(*w, v.(waitItem)) }
func (w *waitQueue) Pop() interface{} {
	n := len(*w)
	tmp := (*w)[n-1]
	*w = (*w)[:n-1]
	return tmp
}

type workItem struct{ finishTime, idx int }
type workQueue []workItem

func (w workQueue) Len() int            { return len(w) }
func (w workQueue) Less(i, j int) bool  { return w[i].finishTime < w[j].finishTime }
func (w workQueue) Swap(i, j int)       { w[i], w[j] = w[j], w[i] }
func (w *workQueue) Push(v interface{}) { *w = append(*w, v.(workItem)) }
func (w *workQueue) Pop() interface{} {
	n := len(*w)
	tmp := (*w)[n-1]
	*w = (*w)[:n-1]
	return tmp
}

func findCrossingTime(n int, k int, time [][]int) int {
	waitLeft := waitQueue{}
	waitRight := waitQueue{}
	workLeft := workQueue{}
	workRight := workQueue{}

	for idx, item := range time {
		heap.Push(&waitLeft, waitItem{item[0] + item[2], idx})
	}

	remain := n
	curTime := 0

	for {
		for len(workLeft) > 0 {
			if workLeft[0].finishTime > curTime {
				break
			}
			tmp := heap.Pop(&workLeft).(workItem)
			heap.Push(&waitLeft, waitItem{time[tmp.idx][0] + time[tmp.idx][2], tmp.idx})
		}
		for len(workRight) > 0 {
			if workRight[0].finishTime > curTime {
				break
			}
			tmp := heap.Pop(&workRight).(workItem)
			heap.Push(&waitRight, waitItem{time[tmp.idx][0] + time[tmp.idx][2], tmp.idx})
		}
		if len(waitRight) > 0 {
			tmp := heap.Pop(&waitRight).(waitItem)
			curTime += time[tmp.idx][2]
			heap.Push(&workLeft, workItem{curTime + time[tmp.idx][3], tmp.idx})
		} else if remain > 0 && len(waitLeft) > 0 {
			remain--
			tmp := heap.Pop(&waitLeft).(waitItem)
			curTime += time[tmp.idx][0]
			heap.Push(&workRight, workItem{curTime + time[tmp.idx][1], tmp.idx})
		} else {
			nextTime := 1 << 30
			if len(workLeft) > 0 {
				nextTime = min(nextTime, workLeft[0].finishTime)
			}
			if len(workRight) > 0 {
				nextTime = min(nextTime, workRight[0].finishTime)
			}
			if nextTime != 1<<30 {
				curTime = max(nextTime, curTime)
			}
		}
		if remain == 0 && len(waitRight) == 0 && len(workRight) == 0 {
			break
		}
	}
	return curTime
}

func min(a int, b int) int {
	if a > b {
		return b
	} else {
		return a
	}
}

func max(a int, b int) int {
	if a > b {
		return a
	} else {
		return b
	}
}
